agent/structs/acl.go

```diff
@@ -140,9 +141,16 @@ type ACLTokenPolicyLink struct {
 	Name string `hash:"ignore"`
 }
 
+// ACLTokenRoleLink is a way to link a Role to a Token. At the storage layer
+// one of ID or BoundName is required. BoundName can only be set during a Login
+// operation when RoleBindingRules are evaluated. If the optional Name is
+// presented during link creation it is resolved to an ID and persisted
+// alongside a snapshot of the current Role's name, but the Name should not be
+// considered a long-term link to the Role, as the ID is used for that.
 type ACLTokenRoleLink struct {
-	ID   string
-	Name string `hash:"ignore"`
+	ID        string `json:",omitempty"`
+	Name      string `json:",omitempty" hash:"ignore"`
+	BoundName string `json:",omitempty"`
```

Comment:
I'm still considering if there is utility in adding a `BoundType string` field here now (with a similar enum added to role binding rules as well).

The current twisty maze of secure intro work is hyperspecific for issuing Tokens that work with Connect. If we wanted to make other variations of magic policies work (agent tokens with restrictive `node:write`) we'd need a way in the role binding rule to indicate an alternate synthetic role be used (`agent` instead of `service` for example), and also on the actual Token->Role link to persist that notation for use a resolve-time.

At resolve-time the pair of `BoundType` and `BoundName` here would be consulted for synthesizing a Role. In the case of `service` it would do what's in this PR (synthetic role with ServiceIdentity.ServiceName set to the role's name). In the case of something like `agent` we'd probably need a little more data on the `ACLTokenRoleLink` structure (like hostname) to be stamped onto the token at Login-time, but then we'd have an avenue to use this plumbing for more than one usage.